app/controllers/concerns/course/lesson_plan/personalization_concern.rb

```diff
@@ -0,0 +1,59 @@
+# frozen_string_literal: true
+module Course::LessonPlan::PersonalizationConcern
+  extend ActiveSupport::Concern
+
+  # Dispatches the call to the correct personalization algorithm
+  # If the algorithm takes too long (e.g. voodoo AI magic), it is responsible for scheduling an async job
+  def update_personalized_timeline_for(course_user, timeline_algorithm = nil)
+    timeline_algorithm ||= course_user.timeline_algorithm
+    send('algorithm_' + timeline_algorithm, course_user)
+  end
+
+  # Fixed timeline: Follow reference timeline
+  # Delete all personal times that are not fixed or submitted
+  def algorithm_fixed(course_user)
+    course_user.reload.personal_times.where(fixed: false, submitted_at: nil).delete_all
+  end
+
+  # TODO: Implement something smarter
+  # For now, copy over reference timeline to personal timeline + fix next 3 items
+  def algorithm_naive(course_user)
+    course_user.reload
+    course_assessments = course_user.course.assessments.
+                         eager_load_reference_times_for(course_user).
+                         eager_load_personal_times_for(course_user).
+                         to_a
+    course_user.transaction do
+      course_assessments.each do |course_assessment|
+        personal_time = course_assessment.eager_loaded_personal_time_for(course_user)
+
+        # Skip committed or submitted items
+        next if personal_time.present? && (personal_time.fixed? || personal_time.submitted_at.present?)
+
+        # Copy over from reference time
+        personal_time = course_assessment.setdefault_personal_time_for(course_user)
```

Comment:
With the suggested modification to `personal_time_for` to init itself to the reference time, this chunk wouldn't be necessary.

Reply:
We can copy over the start/end/bonus_at attributes in setdefault.